reconcile/external_resources/manager.py

```diff
@@ -0,0 +1,330 @@
+                    if reconciliation.resource_hash != state.resource_hash:
+                        reconcile = True
+                        reason = "Resource spec has changed"
+                    elif (
+                        (datetime.now(state.ts.tzinfo) - state.ts).total_seconds()
+                        > reconciliation.module_configuration.reconcile_drift_interval_minutes
+                        * 60
+                    ):
+                        reconcile = True
+                        reason = "Resource drift detection run"
+        elif reconciliation.action == Action.DESTROY:
+            match state.resource_status:
+                case ResourceStatus.CREATED:
+                    reconcile = True
+                    reason = "Resource needs to be deleted"
+                case ResourceStatus.ERROR:
+                    reconcile = True
+                    reason = "Resource status ERROR"
+        if reconcile:
+            logging.info(
+                "Reconciling: Status: [%s], Action: [%s], reason: [%s], key:[%s]",
+                state.resource_status.value,
+                reconciliation.action.value,
+                reason,
+                reconciliation.key,
+            )
+        return reconcile
+
+    def _get_desired_objects_reconciliations(self) -> set[Reconciliation]:
+        r: set[Reconciliation] = set()
+        for key, spec in self.er_inventory.items():
+            module = self.module_inventory.get_from_spec(spec)
+
+            try:
+                resource = self._build_external_resource(spec, self.er_inventory)
+            except ExternalResourceValidationError as e:
+                k = ExternalResourceKey.from_spec(spec)
+                self.errors[k] = e
+                continue
+
+            reconciliation = Reconciliation(
+                key=key,
+                resource_hash=resource.hash(),
+                input=self._serialize_resource_input(resource),
+                action=Action.APPLY,
+                module_configuration=ExternalResourceModuleConfiguration.resolve_configuration(
+                    module, spec
+                ),
+            )
+            r.add(reconciliation)
+        return r
+
+    def _get_deleted_objects_reconciliations(self) -> set[Reconciliation]:
+        desired_keys = set(self.er_inventory.keys())
+        state_resource_keys = set(self.state_mgr.get_all_resource_keys())
+        deleted_keys = state_resource_keys - desired_keys
+        r: set[Reconciliation] = set()
+        for key in deleted_keys:
+            state = self.state_mgr.get_external_resource_state(key)
+            reconciliation = Reconciliation(
+                key=key,
+                resource_hash=state.resource_hash,
+                module_configuration=state.reconciliation.module_configuration,
+                input=state.reconciliation.input,
+                action=Action.DESTROY,
+            )
+            r.add(reconciliation)
+        return r
```

Comment:
can use set comprehension here

```python
        return {
            Reconciliation(
                key=key,
                resource_hash=state.resource_hash,
                module_configuration=state.reconciliation.module_configuration,
                input=state.reconciliation.input,
                action=Action.DESTROY,
            )
            for key in deleted_keys
            if (state := self.state_mgr.get_external_resource_state(key)) is not None
        }
```

if `state` always not None, maybe a Reconciliation builder can make it clear:


```python
        return {
            Reconciliation.build_destroy_action(
                key=key,
                state=self.state_mgr.get_external_resource_state(key),
            )
            for key in deleted_keys
        }
```